doc/developer/platform/architecture-storage.md

```diff
@@ -0,0 +1,131 @@
+the metadata of the collection. Readers use this metadata to know the exact
+time `t_handover` that they should switch from one rendition to another. This
+ensures that even though the collection can transition through multiple
+renditions over time every reader always perceives a definite collection
+advancing in time.
+
+A shard of a rendition is the endpoint where writing happens. Writers insert
+batches of data in a shard by using a write handle on the underlying `persist`
+collection backing the shard.
+
+### Renditions
+
+A rendition represents a version of a collection and is described by its name,
+its `since` frontier, and its `upper` frontier. The `since` frontier is defined
+as the maximum of the `since` frontiers of its shards and its `upper` frontier
+is defined as the minimum of the `upper` frontier of its shards (replace
+max/min with join/meet for partialy ordered times).
+
+#### Rendition metadata
+
+The contents of the collection overall are described by a metadata `persist`
+collection. This rendition metadata collection evolves in time and can be
+thought of as instructions to readers that tells them what data they should
+produce next. It containing updates of this form:
+
+```
+(rendition1, 0, +1)
+// At `t_handover` readers should switch from rendition1 to rendition2
+(rendition1, t_handover, -1)
+(rendition2, t_handover, +1)
+```
+
+Readers hold onto read capabilities for the metadata collection that allows
+STORAGE to know which renditions should be kept around and which ones should be
+garbage collected. In the example above, when all readers downgrade their read
+capability beyond `t_handover` the metadata collection can be consolidated and
+`rendition1` can be deleted.
+
+Readers interpret advances of the medatata frontier as proof that no rendition
+change occured between the previously known frontier and the newly advanced
+frontier. Therefore they can proceed with reading their current rendition for
+all times not beyond the frontier of the metadata collection.
+
+Readers encountering a change in rendition at some handover time `t_handover`
+must produce the necessary diffs such that the resulting collection accumulates
+to the contents of `rendition2` at `t_handover`. A simple strategy for doing so
+is to produce a snapshot of `rendition1` at `t_handover` with all its diffs
+negated and a snapshot of `rendition2` at `t_handover`. More elaborate
+strategies could be implemented where for example STORAGE has precalculated the
+`rendition2 - rendition1` difference collection for readers to use.
+
+### Shards
+
+A shard is part of a rendition and is the endpoint where writers publish new
+data. They can be written to by multiple workers as long as all the workers
+produce identical data (i.e their outputs are definite).
+
+It is an error to have multiple writers write conflicting data into a shard and
+the contents of the collection are undefined if that happens.
```

Comment:
I'm a bit worried about this part, because it seems to have strong implications on the rest of the system -- but I might be mistaken, so hear me out. As I understand it, a shard roughly corresponds to the per-worker state of a DD collection. If there are n Timely workers, we produce n shards, and with AR, we can have several COMPUTE instances, all with n workers writing the same n shards.

For COMPUTE, we want to allow AR between instances of different configuration, meaning that they could have a different number of Timely workers but still be part of the same AR group. How would STORAGE handle writing from differently-sized Timely instances? Would it result in the AR group writing to different renditions, which would coexist in parallel?

Reply:
> How would STORAGE handle writing from differently-sized Timely instances? Would it result in the AR group writing to different renditions, which would coexist in parallel?

Yes, that's what would happen. To be more specific, STORAGE doesn't particularly about a 1-1 mapping between a STORAGE shard and a timely worker. We could for example have an AR setup where the number of workers between the clusters are different with the constraint being that for a given STORAGE shard there is only one worker responsible for it and we have an exchange as the last step of the process.